Reattempt pod creation in the face of ResourceQuota errors

sbwsg opened the pull request:
# Changes

Fixes https://github.com/tektoncd/pipeline/issues/734

## TLDR

[TaskRuns](https://github.com/tektoncd/pipeline/tree/master/docs/taskruns.md) used to fail immediately if their pods hit the ceiling of a [ResourceQuota](https://kubernetes.io/docs/concepts/policy/resource-quotas/). Now they don't fail immediately and instead reattempt creating the pod until success or timeout.

## Long Version:

TaskRuns create pods. Pods created in a namespace with ResourceQuota can be rejected on the basis of their resource requirements. This rejection manifests in the TaskRun reconciler as an error from the
createPod() function.

Prior to this commit all errors from createPod() were considered fatal and the associated TaskRun would be marked as failed. This commit introduces a process for reattempting pod creation in the face of ResourceQuota errors.

When a ResourceQuota error is encountered, the TR reconciler now performs the following work:

- The TaskRun is marked as Succeeded/Unknown with reason ExceededResourceQuota and a message including the number of reattempts that have been made
- A reconcile is scheduled for the TaskRun using a backoff w/ jitter strategy

Pod creation will be continually reattmpted until the ResourceQuota errors stop or the TaskRun times out.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] ~~Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)~~
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

<h1>Screenshots</h1>

<details>
<summary>Before</summary>

TaskRuns that hit a ResourceQuota ceiling immediately fail out:

![2019-05-24_13-23-49](https://user-images.githubusercontent.com/34253460/58346496-1a35c080-7e29-11e9-889b-ad03c47aa046.png)
</details>

<details>
<summary>After</summary>

TaskRuns that hit a ResourceQuota are placed into a Succeeded/Unknown state and retry their pod creation until success or timeout.

![2019-05-24_13-20-49](https://user-images.githubusercontent.com/34253460/58346511-24f05580-7e29-11e9-8600-1dddb349f992.png)

</details>

# Release Notes

```
Task pods that hit resource limits imposed by ResourceQuota objects in a namespace will no longer fail their TaskRun immediately. Pod creation will be repeatedly tried until successfully started or the TaskRun times out.
```

abayer commented:
/lgtm

dlorenc commented on pkg/reconciler/timeout_handler_test.go:
```diff
@@ -259,3 +259,80 @@ func TestWithNoFunc(t *testing.T) {
+		}
+		endTime := GetTaskRunEndTime(taskRun)
+		if !endTime.Equal(tc.expectedEndTime) {
+			t.Fatalf("testcase %d expected %q received %q", i+1, tc.expectedEndTime.Format(time.StampNano), endTime.Format(time.StampNano))
```
nit: if you make these subtests (use t.Run and t.Fail) instead of t.Fatal, then the test runner will run them all even if the first one fails.

dlorenc commented on pkg/reconciler/timeout_handler_test.go:
```diff
@@ -259,3 +259,80 @@ func TestWithNoFunc(t *testing.T) {
+		expectedEndTime time.Time
+	}{
+		{
+			startTime:       timeNow,
```
style nit: if you add a description to each case as well, you can pass that to t.Run like I explained below. Then the test cases are a bit easier to read and failures are easier to debug.

dlorenc commented on pkg/reconciler/timeout_handler_test.go:
```diff
@@ -259,3 +259,80 @@ func TestWithNoFunc(t *testing.T) {
+	c, _ := test.SeedTestData(t, d)
+	observer, _ := observer.New(zap.InfoLevel)
+	testHandler := NewTimeoutHandler(c.Kube, c.Pipeline, stopCh, zap.New(observer).Sugar())
+	dur := 50 * time.Millisecond
```
where does this duration come from? Is it just random?

sbwsg replied:
Renamed variable to be more clear on its purpose here - it's the timer's duration. Added another variable beneath this one to document the timer's failure case more clearly.

dlorenc commented on pkg/reconciler/timeout_handler_test.go:
```diff
@@ -259,3 +259,80 @@ func TestWithNoFunc(t *testing.T) {
+	}
+	testHandler.SetTaskRunCallbackFunc(callback)
+	testHandler.SetTaskRunTimer(taskRun, dur)
+	<-time.After(100 * time.Millisecond)
```
Hmm, I always get a weird feeling when I see hardcoded sleeps in tests. What about another channel you wait on with a select (and a timeout like this) that you write to in the callback?

sbwsg replied:
Nice, updated to a `select {}` that relies on a done channel and an expiration timer.

vdemeester commented:
/lgtm

bobcatfish commented on pkg/reconciler/v1alpha1/taskrun/taskrun.go:
```diff
@@ -320,12 +328,12 @@ func (c *Reconciler) reconcile(ctx context.Context, tr *v1alpha1.TaskRun) error
 			})
 			c.Recorder.Eventf(tr, corev1.EventTypeWarning, "BuildCreationFailed", "Failed to create build pod %q: %v", tr.Name, err)
-			c.Logger.Errorf("Failed to create build pod for task %q :%v", err, tr.Name)
+			c.Logger.Errorf("Failed to create build pod for task %q: %v", tr.Name, err)
```
what do you think about moving the logic around determining the status of the Run out of this function and into a separate function that could have its own tests?

bobcatfish commented on pkg/reconciler/v1alpha1/taskrun/taskrun.go:
```diff
@@ -302,15 +302,23 @@ func (c *Reconciler) reconcile(ctx context.Context, tr *v1alpha1.TaskRun) error
 			if isExceededResourceQuotaError(err) {
 				reason = reasonExceededResourceQuota
-				msg = getExceededResourcesMessage(tr)
+				count, deadline, isNewBackoff := c.timeoutHandler.GetBackoff(tr)
```
i have some total nitpicks about these variable names - mostly cuz i had to read this several times before i understood why they were named the way they were. This is basically what i thought they meant:

- count: i really wasnt sure, maybe the total count of backoffs the system had attempted ever?
- deadline: the time by which this taskrun must succesfully complete or we give up on it
- isNewBackoff: whether or not the "backoff" structure was newly created, i.e. whether or not we've started backing off for this item

This is what I think these _actually_ are:

- count: number of attempts for this taskrun
- deadline: time before the next attempt
- isNewBackoff: whether or not we are in the middle of backing off for this task run

So maybe some alternative names (not sure these are great, feel free to change/ignore):

- count -> numAttempts
- deadline -> nextAttemptTime
- isNewBackoff -> currentlyBackingOff

(this is super minor)

sbwsg replied:
Agreed, the original names are not obvious.  Will fix.

bobcatfish commented on pkg/reconciler/v1alpha1/taskrun/taskrun.go:
```diff
@@ -302,15 +302,23 @@ func (c *Reconciler) reconcile(ctx context.Context, tr *v1alpha1.TaskRun) error
+				if isNewBackoff {
+					taskRunEndTime := reconciler.GetTaskRunEndTime(tr)
+					if deadline.After(taskRunEndTime) {
+						deadline = taskRunEndTime
```
since `GetBackoff` takes the entire taskRun as a param (not that it should - maybe it shouldnt), I'm kind of surprised it doesn't know how to determine if the deadline is past the end of the execution time :thinking:

sbwsg commented on pkg/reconciler/timeout_handler.go:
```diff
@@ -101,6 +119,47 @@ func GetTimeout(d *metav1.Duration) time.Duration {
+	b.count += 1
+	b.deadline = time.Now().Add(backoffDuration(b.count))
+	t.backoffs[runObj.GetRunKey()] = b
+	return b.count, b.deadline, true
```
The `backoff` struct felt to me like an implementation detail of the internal state of the backoff-tracking mechanism and I didn't sense a strong reason to expose a new public type in the package's interface.  I don't feel strongly about this and am happy to change it but am curious what might be the benefit of making it public?

bobcatfish replied:
> but am curious what might be the benefit of making it public?

The only one I can think of is simplifying the signature of the function a bit - since you already have a structure!

It reminded me of a pattern I've often seen where a function takes too many parameters or returns too many things, and this is often a sign that another abstraction can be added that would simplify the interface (this website talks about it a bit: https://refactoring.guru/smells/long-parameter-list disclaimer I have never seen this website before haha) - and it seemed like what was happening here was doing the opposite

Just a super minor thing tho!

sbwsg commented on pkg/reconciler/timeout_handler.go:
```diff
@@ -101,6 +119,47 @@ func GetTimeout(d *metav1.Duration) time.Duration {
+	return b.count, b.deadline, true
+}
+
+func backoffDuration(count uint) time.Duration {
```
oh interesting, i hadn't considered injecting my own `rand.Intn` and gave up testing this func because i couldn't think of a way to get deterministic results.  Cheers, I'll give this a shot!

sbwsg replied:
I've extracted the jitter func, set it to `rand.Intn()` by default and added testcases for this.

sbwsg commented on pkg/reconciler/timeout_handler.go:
```diff
@@ -195,3 +254,35 @@ func (t *TimeoutSet) waitRun(runObj StatusKey, timeout time.Duration, startTime
+	t.setTimer(tr, d, callback)
+}
+
+func (t *TimeoutSet) setTimer(runObj StatusKey, timeout time.Duration, callback func(interface{})) {
```
Totally, will make the change.  I hesitated to do this initially because the log messages are slightly different.

sbwsg replied:
I've merged the timer logic between these two funcs.

bobcatfish commented on pkg/reconciler/timeout_handler.go:
```diff
@@ -12,19 +14,30 @@ import (
+	// NumAttempts reflects the number of times a given StatusKey has been delayed
+	NumAttempts uint
+	// NextAttempt is the point in time at which this backoff expires
+	NextAttempt time.Time
```
ah nice, i like the new names!

bobcatfish commented on pkg/reconciler/timeout_handler_test.go:
```diff
@@ -259,3 +259,108 @@ func TestWithNoFunc(t *testing.T) {
+		t.Run(tc.description, func(t *testing.T) {
+			receivedDuration := GetTimeout(tc.inputDuration)
+			if receivedDuration != tc.expectedDuration {
+				t.Errorf("expected %q received %q", tc.expectedDuration.String(), receivedDuration.String())
```
nice! i think the test is super clear this way :D

bobcatfish commented on pkg/reconciler/timeout_handler_test.go:
```diff
@@ -259,3 +259,108 @@ func TestWithNoFunc(t *testing.T) {
+	stopCh := make(chan struct{})
+	c, _ := test.SeedTestData(t, d)
+	observer, _ := observer.New(zap.InfoLevel)
+	testHandler := NewTimeoutHandler(c.Kube, c.Pipeline, stopCh, zap.New(observer).Sugar())
```
i just realized something but _this is totally just scope creep_ so I totally understand if you want to ignore me XD

I looked at this test and I was like "why does it need a kube client and a pipeline client"? and after looking into it it looks like those clients only exist in the timeout handler [to support this one call to `CheckTimeouts`](https://github.com/tektoncd/pipeline/blob/11d1ca8cb7a019dda75ed7deea4764d5cdc68aaa/cmd/controller/main.go#L144) - which means that instead of passing the clients into the timeout handler, the `CheckTimeouts` function could take them as arguments :D which if I'm right would simplify the setup required for this test

anyway feel free to ignore this  XD

bobcatfish commented on pkg/reconciler/timeout_handler_test.go:
```diff
@@ -259,3 +259,108 @@ func TestWithNoFunc(t *testing.T) {
+	timerFailDeadline := 100 * time.Millisecond
+	doneCh := make(chan struct{})
+	callback := func(_ interface{}) {
+		close(doneCh)
```
you could also have this function set a boolean or something that would could use to assert that the function was called or not called as expected

sbwsg replied:
I initially implemented it using a bool that the callback mutated but this felt a little icky due to the inline sleep:

```go
done := false
testHandler.SetTaskRunCallbackFunc(func(_ interface{}) { done = true })
go testHandler.SetTaskRunTimer(taskRun, timerDuration)
<-time.After(timerFailDeadline) // ick?
if !done {
  t.Errorf("timer did not execute task run callback func within expected time")
}
```

I kind of like how the `select` approach races the two channels against each other.  Full credit to @dlorenc (https://github.com/tektoncd/pipeline/pull/905#discussion_r287564205) for pointing this out.

I may have completely misinterpreted either or both approaches though so lmk if there's something I'm missing.

bobcatfish commented on pkg/reconciler/timeout_handler_test.go:
```diff
@@ -259,3 +259,108 @@ func TestWithNoFunc(t *testing.T) {
+	packageJitterFunc := jitterFunc
+	for _, tc := range testcases {
+		t.Run(tc.description, func(t *testing.T) {
+			jitterFunc = tc.jitterFunc
```
super minor: you could do this assignment before the for loop (i.e. doesnt need to happen for every test)

vdemeester replied:
@bobcatfish I am guessing this is resolved as the `jitterFunc` comes from the test cases (so got to be `packageJitterFunc`)

vdemeester commented on pkg/reconciler/timeout_handler_test.go:
```diff
@@ -259,3 +259,108 @@ func TestWithNoFunc(t *testing.T) {
+	for _, tc := range testcases {
+		t.Run(tc.description, func(t *testing.T) {
+			jitterFunc = tc.jitterFunc
+			result := backoffDuration(tc.inputCount)
```
I wonder if we could pass the `jitterFunc as an argument to `backoffDuration` instead. This would remove the need to set (or have) the `jitterFunc` global variable (just for the test) :angel:

vdemeester replied:
```go
# in code
backoffDuration(b.NumAttempts, rand.Intn)
# in tests
backoffDuration(tc.inputCount, tc.jitterFunc)
```

sbwsg commented:
@bobcatfish @vdemeester thanks for the feedback so far.  I've updated `jitterFunc` to be an arg passed to `backoffDuration()`, moved taskrun status -related code to its own method (w/ UTs) and modified `CheckTimeouts()` to take clientsets as arguments.  That last change required some small refactoring in `controller/main.go` and the pipelinerun / taskrun reconcilers.

bobcatfish commented on pkg/reconciler/timeout_handler.go:
```diff
@@ -122,22 +181,22 @@ func (t *TimeoutSet) checkPipelineRunTimeouts(namespace string) {
 // taskrun/pipelinerun timeout functions
-func (t *TimeoutSet) CheckTimeouts() {
-	namespaces, err := t.kubeclientset.CoreV1().Namespaces().List(metav1.ListOptions{})
+func (t *TimeoutSet) CheckTimeouts(kubeclientset kubernetes.Interface, pipelineclientset clientset.Interface) {
```
niiiiice thanks for this cleanup with the clients!!! :heart_eyes: :heart_eyes: :heart_eyes:

bobcatfish commented:
Looks awesome!! Thanks @sbwsg :D and thanks for the clean up too ^.^

![image](https://user-images.githubusercontent.com/432502/58581976-9c492f00-8205-11e9-8c37-0844b96072ad.png)

/lgtm

vdemeester commented:
ah @sbwsg, needs a rebase :bowing_man:

sbwsg commented:
:crossed_fingers:

vdemeester commented:
/lgtm
